librz/main/rz-diff.c

```diff
@@ -1,1305 +1,1398 @@
-			       "0 1 2 3 4 5 6 7 8 9 A B C D E F 0123456789ABCDEF\n");
-		break;
-	default:
-		eprintf("Invalid column width\n");
-		return;
+
+	rz_config_set_b(core->config, "scr.interactive", false);
+	rz_config_set_b(core->config, "cfg.debug", false);
+	core->print->scr_prompt = false;
+
+#if __WINDOWS__
+	char *winf = rz_acp_to_utf8(filename);
+	cfile = rz_core_file_open(core, winf, 0, 0);
+	free(winf);
+#else
+	cfile = rz_core_file_open(core, filename, 0, 0);
+#endif
+	if (!cfile) {
+		rz_diff_error("cannot open file '%s'\n", filename);
+		goto rz_diff_load_file_with_core_fail;
 	}
-	rz_cons_break_push(NULL, NULL);
-	for (i = 0; i < sz; i += w) {
-		if (rz_cons_is_breaked()) {
-			break;
-		}
-		if (i + w >= sz) {
-			pad = w - sz + i;
-			w = sz - i;
-		}
-		bool eq = !memcmp(a + i, b + i, w);
-		if (eq) {
-			ctx--;
-			if (ctx == -1) {
-				rz_cons_printf("...\n");
-				continue;
-			}
-			if (ctx < 0) {
-				ctx = -1;
-				continue;
-			}
-		} else {
-			ctx = DUMP_CONTEXT;
-		}
-		rz_cons_printf(eq ? Color_GREEN : Color_RED);
-		rz_cons_printf("0x%08x%c ", i, eq ? ' ' : '!');
-		rz_cons_printf(Color_RESET);
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%02x", a[i + j]);
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf("  ");
-		}
-		rz_cons_printf(" ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%c", IS_PRINTABLE(a[i + j]) ? a[i + j] : '.');
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf(" ");
-		}
-		rz_cons_printf("   ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%02x", b[i + j]);
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		for (j = 0; j < pad; j++) {
-			rz_cons_printf("  ");
-		}
-		rz_cons_printf(" ");
-		for (j = 0; j < w; j++) {
-			bool eq2 = a[i + j] == b[i + j];
-			if (!eq) {
-				rz_cons_printf(eq2 ? Color_GREEN : Color_RED);
-			}
-			rz_cons_printf("%c", IS_PRINTABLE(b[i + j]) ? b[i + j] : '.');
-			if (!eq) {
-				rz_cons_printf(Color_RESET);
-			}
-		}
-		rz_cons_printf("\n");
-		rz_cons_flush();
+
+	if (!rz_core_bin_load(core, NULL, UT64_MAX)) {
+		rz_diff_error("cannot load file '%s'\n", filename);
+		goto rz_diff_load_file_with_core_fail;
 	}
-	rz_cons_break_end();
-	rz_cons_printf("\n" Color_RESET);
-	rz_cons_flush();
-	if (as != bs) {
-		rz_cons_printf("...\n");
+
+	if (!rz_core_bin_update_arch_bits(core)) {
+		rz_diff_error("cannot set architecture with bits\n");
+		goto rz_diff_load_file_with_core_fail;
+	}
+
+	if (rz_list_empty(rz_bin_get_sections(core->bin))) {
+		rz_config_set_i(core->config, "io.va", false);
```

Comment:
What is the exact reason for this? I think you want to query for whether the file has maps, rather than sections. To avoid using the deprecated global `rz_bin_get_*` functions, take the binfile from the `cfile`.
The same goes for all the other bin functions used below, try to pass through an `RzBinFile *` or `RzBinObject *`.

Reply:
how i do this?

Reply:
If you want to check if the binfile has bin-maps, check whether `cfile->binfiles[0]->o->maps` is not empty after loading. These are the maps that instruct core how to create io maps.
But then again, the question is what this is actually trying to achieve. I guess if there is no virtual mapping for the file, it should just use file file as-is. But iirc in such a case (which is when the bin-maps mentioned above are empty), core already maps the entire file as a single `RzIOMap` so switching `io.va` to false would change some internals but you would get the same results. So I think this needs some clarification/docs and tests for both cases.